File: prime_sieve/sieve.py
#!/usr/bin/env python3
import math

def sieve_of_eratosthenes(n):
    """Find all primes up to n using Sieve of Eratosthenes"""
    if n < 2:
        return []
    
    # Create boolean array, initially all True
    primes = [True] * (n + 1)
    primes[0] = primes[1] = False
    
    # Sieve algorithm
    for i in range(2, int(math.sqrt(n)) + 1):
        if primes[i]:
            for j in range(i * i, n + 1, i):
                primes[j] = False
    
    # Collect prime numbers
    result = []
    for i in range(2, n + 1):
        if primes[i]:
            result.append(i)
    
    return result

def main():
    limit = 10000000  # 10 million
    primes = sieve_of_eratosthenes(limit)
    
    print(f"Found {len(primes)} primes up to {limit}")
    print(f"First 10 primes: {primes[:10]}")
    print(f"Last 10 primes: {primes[-10:]}")

if __name__ == "__main__":
    main()

File: prime_sieve/sieve.php
<?php

function sieveOfEratosthenes($n) {
    if ($n < 2) {
        return array();
    }
    
    // Create boolean array
    $primes = array_fill(0, $n + 1, true);
    $primes[0] = $primes[1] = false;
    
    // Sieve algorithm
    for ($i = 2; $i * $i <= $n; $i++) {
        if ($primes[$i]) {
            for ($j = $i * $i; $j <= $n; $j += $i) {
                $primes[$j] = false;
            }
        }
    }
    
    // Collect prime numbers
    $result = array();
    for ($i = 2; $i <= $n; $i++) {
        if ($primes[$i]) {
            $result[] = $i;
        }
    }
    
    return $result;
}

function main() {
    $limit = 1000000;  // 1 million (reduced for memory constraints)
    $primes = sieveOfEratosthenes($limit);
    
    echo "Found " . count($primes) . " primes up to " . $limit . "\n";
    echo "First 10 primes: " . implode(" ", array_slice($primes, 0, 10)) . "\n";
    echo "Last 10 primes: " . implode(" ", array_slice($primes, -10)) . "\n";
}

main();

?>
